pkg/controller/servicemesh/webhook/controller.go

```diff
@@ -0,0 +1,247 @@
+	err = c.Watch(&source.Kind{Type: &v1beta1.MutatingWebhookConfiguration{}}, &handler.EnqueueRequestForObject{}, predicate.Funcs{
+		CreateFunc: func(event event.CreateEvent) bool {
+			return strings.HasPrefix(event.Meta.GetName(), sidecarInjectorWebhookConfigNamePrefix)
+		},
+		UpdateFunc: func(event event.UpdateEvent) bool {
+			return strings.HasPrefix(event.MetaNew.GetName(), sidecarInjectorWebhookConfigNamePrefix)
+		},
+		// deletion and generic events don't interest us
+		DeleteFunc: func(event event.DeleteEvent) bool {
+			return false
+		},
+		GenericFunc: func(event event.GenericEvent) bool {
+			return false
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	// Watch ValidatingWebhookConfigurations
+	err = c.Watch(&source.Kind{Type: &v1beta1.ValidatingWebhookConfiguration{}}, &handler.EnqueueRequestForObject{}, predicate.Funcs{
+		CreateFunc: func(event event.CreateEvent) bool {
+			return strings.HasPrefix(event.Meta.GetName(), galleyWebhookConfigNamePrefix)
+		},
+		UpdateFunc: func(event event.UpdateEvent) bool {
+			return strings.HasPrefix(event.MetaNew.GetName(), galleyWebhookConfigNamePrefix)
+		},
+		// deletion and generic events don't interest us
+		DeleteFunc: func(event event.DeleteEvent) bool {
+			return false
+		},
+		GenericFunc: func(event event.GenericEvent) bool {
+			return false
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// reconciles webhook configurations
+type reconciler struct {
+	common.ResourceManager
+	scheme *runtime.Scheme
+}
+
+// Reconcile updates ClientConfigs of webhook configurations to contain the CABundle
+// from the respective Istio SA secret
+func (r *reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	logger := log.WithValues("WebhookConfig", request.Name)
+	updated := false
+	var config runtime.Object
+
+	if strings.HasPrefix(request.Name, sidecarInjectorWebhookConfigNamePrefix) {
+		logger.Info("reconciling MutatingWebhookConfiguration")
+		// get current webhook config
+		currentConfig := &v1beta1.MutatingWebhookConfiguration{}
+		err := r.Client.Get(context.TODO(), request.NamespacedName, currentConfig)
+		if err != nil {
+			log.Info("MutatingWebhookConfiguration does not exist yet. No action taken")
+			return reconcile.Result{}, nil
+		}
+		namespace := request.Name[len(sidecarInjectorWebhookConfigNamePrefix):]
+		// retrieve root certificate from secret
+		caRoot, err := r.getCABundle(namespace, sidecarInjectorServiceAccountSecretName)
+		if err != nil {
+			logger.Info("could not get secret: " + err.Error())
+			return reconcile.Result{}, nil
+		}
+		// update caBundle if it doesn't match what's in the secret
+		newConfig := currentConfig.DeepCopyObject().(*v1beta1.MutatingWebhookConfiguration)
+		for i := range newConfig.Webhooks {
+			updated = injectCABundle(&newConfig.Webhooks[i].ClientConfig, caRoot) || updated
+		}
+		config = newConfig
+	} else if strings.HasPrefix(request.Name, galleyWebhookConfigNamePrefix) {
```

Comment:
Ugh, so this reconciler reconciles two different kinds of objects. I haven't seen this pattern in any existing controller. IMHO, this really should be two different controllers.

Reply:
Alright then, let's go there.